pandas/core/groupby.py

```diff
@@ -3740,9 +3740,21 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        indexer = self.grouper.result_index.get_indexer(index)
+        result = result.reindex(index=indexer, copy=False)
+        result.index = range(len(result))
+        # Update values of grouper columns with index levels
+        for i in range(index.nlevels):
```

Comment:
I don't know how to do it in another way. I wanted to get this:
```python
df = pd.DataFrame({'cat': pd.Categorical([1, 2, 2], [1, 2, 3]), 'A': [10, 11, 11], 'B': [101, 102, 103]})

df.groupby(['cat', 'A'], as_index=False).sum()
Out[6]: 
   cat   A      B
0    1  10  101.0
1    1  11    NaN
2    2  10    NaN
3    2  11  205.0
4    3  10    NaN
5    3  11    NaN

# for comparison, with as_index=True:
df.groupby(['cat', 'A']).sum()
Out[7]: 
            B
cat A        
1   10  101.0
    11    NaN
2   10    NaN
    11  205.0
3   10    NaN
    11    NaN
```
Without lines 3754-3756, the output would be:
```python
# df.groupby(['cat', 'A'], as_index=False).sum()
Out: 
      A    C      B
0  10.0  1.0  101.0
1   NaN  NaN    NaN
2   NaN  NaN    NaN
3   NaN  NaN    NaN
4  11.0  2.0  205.0
5   NaN  NaN    NaN
```
In such a case, expanding the data frame doesn't make sense to me. 

But if we don't want to expand the categorical column, ie if the following output is enough:
```python
# df.groupby(['cat', 'A'], as_index=False).sum()
Out: 
   cat   A      B
0    1  10  101
1    2  11  205
```
then the whole fix gets very simple (just return the original `result` without any changes).